File: Week_03/G20190343020270/Generate Parentheses.py
# 22. Generate Parentheses
from collections import deque

class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def backtrack(s = '', left = 0, right = 0):
            if left == n and right == n:
                res.append(s)
                # return
            if left < n:
                backtrack(s+'(', left+1, right)
            if right < left:
                backtrack(s+')', left, right+1)
            
        res = []
        backtrack()
        return res

if __name__ == "__main__":
    solution = Solution()
    print(solution.generateParenthesis(3))

File: Week_03/G20190343020104/java/leetcode/editor/cn/GenerateParentheses.java
//给出 n 代表生成括号的对数，请你写出一个函数，使其能够生成所有可能的并且有效的括号组合。 
//
// 例如，给出 n = 3，生成结果为： 
//
// [
//  "((()))",
//  "(()())",
//  "(())()",
//  "()(())",
//  "()()()"
//]
// 
// Related Topics 字符串 回溯算法

package leetcode.editor.cn;

import java.util.ArrayList;
import java.util.List;

public class GenerateParentheses {
    public static void main(String[] args) {
        Solution solution = new GenerateParentheses().new Solution();
        System.out.println(solution.generateParenthesis(2));
    }


    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        List<String> list = new ArrayList<>();

        public List<String> generateParenthesis(int n) {
            _generate(0, 0, 2, "");
            return list;
        }

        private void _generate(int left, int right, int n, String s) {
            //terminator
            if (left == n && right == n) {
                list.add(s);
                return;
            }
            //process

            //drill down
            if (left < n) _generate(left + 1, right, n, s + "(");
            if (left > right) _generate(left, right + 1, n, s + ")");
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}